Learner 7611 - Make URLs of all video encodings available in Blocks API

jawad-khan opened the pull request:
Make URLs of all video encodings available in Blocks API.

student_view_data in VideoBlock expects profiles in context, so I used this option to include all encodings in blocks api.
I tried to go through param option but architecture didn't allow, approach can be seen in commits.

nasthagiri commented on lms/djangoapps/course_api/blocks/transformers/student_view.py:
```diff
@@ -14,6 +14,11 @@ class StudentViewTransformer(BlockStructureTransformer):
     STUDENT_VIEW_DATA = 'student_view_data'
     STUDENT_VIEW_MULTI_DEVICE = 'student_view_multi_device'
+    STUDENT_VIEW_DATA_CONTEXT = {
+        'video': {
```
Can this be implemented without infiltrating the generic core API with block-specific details?
Why not have the video-block know which contexts it should be sending back? That is, just include `STUDENT_VIEW_DATA_CONTEXT[video]` within the video block's implementation.

jawad-khan replied:
@nasthagiri You mean making [VideoBlock](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/video_module/video_module.py#L1084) send all encodings by default and then transforming it in [VideoBlockURLTransformer](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/transformers/video_urls.py#L15)?

nasthagiri replied:
Yes, that works.
Also, VideoBlockURLTransformer probably should live outside of the `course_api` Django app - in Video's own Django app/folder, along with `video_module.py`.

jawad-khan replied:
@nasthagiri I have moved STUDENT_VIEW_DATA_CONTEXT[video] inside video_module.py. VideoBlockURLTransformer contains blocks api related attributes like student_view_data and StudentViewTransformer. Also it isn't used by any package outside of course_api.

jawad-khan commented:
@nasthagiri Please have a look.

nasthagiri commented:
@jawad-khan Am I reading this right? Did you simplify this change down to a single line of code? Adding more video profiles to the API response?

Are there any concerns about increasing the payload response size? It seems from the commit history that you were attempting this path, but ran into issues?

Is the approach now to move forward with this PR and then update if there's a performance concern?

jawad-khan commented:
@nasthagiri yes I have simplified it to a single line of code and added more video profiles to the api response. 
From commit history you can see I started by trying to make it flexible through api params in blocks api but couldn't achieve this because of current architecture.
Then I moved these video profiles in student transformer but you suggested to move this logic in video block. I still think we should move this in student transformer which eventually will have less impacts on all other apis. 
we can have a short discussion on this if you want.
Currently, yes I want to move forward with this PR and will get back to it if there are performance issues.

nasthagiri commented:
_> we should move this in student transformer which eventually will have less impact on all other apis_

If you can do this without breaking architectural boundaries between the core API implementation and the details of the video XBlock, then that would be great. There is probably a way to do so via the `context` param.

> Currently, yes I want to move forward with this PR and will get back to it if there are performance issues.

What are your plans for monitoring the performance issues?

nasthagiri reviewed (approved):
Reminder to:
* squash commits
* use [conventional commit](https://github.com/edx/open-edx-proposals/pull/189) style naming for the commit message and the PR title
* ensure you have a plan for monitoring and addressing performance degradation

Also, note that you'll want to add/modify a unit test for this in [VideoBlockStudentViewDataTestCase](https://github.com/edx/edx-platform/blob/38ee797cfe1c5279559349c17ead961e1811548a/common/lib/xmodule/xmodule/tests/test_video.py#L826).

jawad-khan commented:
> If you can do this without breaking architectural boundaries between the core API implementation and the details of the video XBlock, then that would be great. There is probably a way to do so via the `context` param.
Doing this through context param is the best solution, can you please guide me how can I achieve this?
